File: 프로그래머스/3/70130. 스타 수열/스타 수열.cpp
#include <string>
#include <vector>
#include <map>
#include <queue>
using namespace std;
int solution(vector<int> a) {
    int answer = 0;
    map<int,int>ma;
    for(int i=0;i<a.size();i++){
        ma[a[i]]++;
    }
    priority_queue<pair<int,int>> pq;
    for(auto&i:ma){
        pq.push({i.second,i.first});
    }
    while(!pq.empty()){
        if(pq.top().first*2<answer){
            break;
        }
        int k=pq.top().second;
        pq.pop();
        int cnt=0;
        for(int i=0;i<a.size()-1;i++){
            if((a[i]==k&&a[i+1]!=k)||(a[i]!=k&&a[i+1]==k)){
                cnt+=2;
                i++;
            }
        }
        answer=max(answer,cnt);
    }
    
    return answer;
}

File: 프로그래머스/3/70130. 스타 수열/스타 수열.java
import java.util.*;

class Solution {
	public int solution(int[] a) {
		int answer = 0;
		Map<Integer, Integer> map = new HashMap<>();
		for (int i = 0; i < a.length; i++) {
			map.put(a[i], map.getOrDefault(a[i], 0) + 1);
		}
		Queue<Pair> pq = new PriorityQueue<>();
		for (var v: map.entrySet()) {
			pq.add(new Pair(v.getValue(), v.getKey()));
		}
		while (!pq.isEmpty()) {
			if (pq.peek().first * 2 < answer) {
				break;
			}
			int k = pq.peek().second;
			pq.poll();

			int cnt = 0;
			for (int i = 0; i < a.length - 1; i++) {
				if ((a[i] == k && a[i + 1] != k) || (a[i] != k && a[i + 1] == k)) {
					cnt += 2;
					i++;
				}
			}
			answer = Math.max(answer, cnt);
		}

		return answer;
	}

	class Pair implements Comparable<Pair> {
		public int first;
		public int second;

		public Pair(int first, int second) {
			this.first = first;
			this.second = second;
		}

		@Override
		public int compareTo(Pair p) {
			if (this.first > p.first || (this.first == p.first && this.second > p.second)) {
				return -1;
			} else {
				return 1;
			}
		}
	}
}